### PR TITLE
feat: int division operator

### LINKED ISSUE #2426: Introduce integer division operator `//`
### What's up?

DuckDB seems to introduce a new division operator `/` and  the current division operator `/` will be moved to `//` (duckdb/duckdb#7082).

I'm wondering if the same thing could be done with PRQL.

In my opinion, PRQL should not do the division itself as it currently does, but should convert `//` to `/` for targets that do not have two division operators.

> Integer Division: SQLite, Postgres, SQL Server
> Floating Point Division: MySQL, SparkSQL, BigQuery, Snowflake, ClickHouse

Note that My SQL seems to require the `DEV` operator for integer division.

### PR DESCRIPTION
**Pull request by aljazerzen**
Closes #2426

Also makes sure that float division is always producing floats.
This is done by casting to float before division. This can be removed
when match is implemented and we can check types and use only the
appropriate operation.

### CODE REVIEW

File: prql-compiler/src/sql/std.sql.prql [hunk 1]
```diff
@@ -38,7 +38,8 @@ let read_parquet = source -> s"read_parquet({source})"
 let read_csv = source -> s"read_csv_auto({source})"
 
 let mul = l r -> null
-let div = l r -> null
+let div_i = l r -> s"ROUND({l} / {r} - 0.5)"
+let div_f = l r -> s"(CAST({l} AS FLOAT) / {r})"
```
Comment (eitsupi) [inline prql-compiler/src/sql/std.sql.prql]: I am not sure that these are always correct.

- If `ROUND` is "Bankers' rounding", `ROUND(x - 0.5)` would not be `FROOR(x)`.
- If `l` is a type like FLOAT64, I think `CAST(l AS FLOAT)` is not correct. (See PEP 238 cited in this comment https://github.com/PRQL/prql/issues/2426#issuecomment-1554694777)

Comment (aljazerzen) [reply prql-compiler/src/sql/std.sql.prql]: > If ROUND is "Bankers' rounding", ROUND(x - 0.5) would not be FLOOR(x).

I used FLOOR before, but it was giving wrong answers for `-13 / 5`. I'll run tests on all databases locally now.

---

>If l is a type like FLOAT64, I think CAST(l AS FLOAT) is not correct.

Their argument is that it would not work for complex numbers. I don't think we should concern ourselves with that yet.

Comment (aljazerzen) [reply prql-compiler/src/sql/std.sql.prql]: > If ROUND is "Bankers' rounding", ROUND(x - 0.5) would not be FLOOR(x).

I used FLOOR before, but it was giving wrong answers for `-13 / 5`. I'll run tests on all databases locally now.

---

>If l is a type like FLOAT64, I think CAST(l AS FLOAT) is not correct.

Their argument is that it would not work for complex numbers. I don't think we should concern ourselves with that yet.

Comment (aljazerzen) [reply prql-compiler/src/sql/std.sql.prql]: On MySQL, `x * 1.0` somehow produces different results than `CAST(x AS FLOAT)`. I think it's converting for decimal instead of float.

File: prql-compiler/src/sql/std.sql.prql [hunk 2]
```diff
@@ -38,8 +38,9 @@ let read_parquet = source -> s"read_parquet({source})"
 let read_csv = source -> s"read_csv_auto({source})"
 
 let mul = l r -> null
-let div = l r -> null
-let mod = l r -> null
+let div_i = l r -> s"FLOOR(FLOOR(ABS({l} / {r})) * SIGN({l} / {r}))"
+let div_f = l r -> s"({l} * 1.0 / {r})"
```
Comment (max-sixty) [inline prql-compiler/src/sql/std.sql.prql:42]: One downside of s-strings is that if `l` is `5 + 5`, then it doesn't understand precedence, and becomes `(5 + 5 * 1.0 / 2)`, so the `/ 2` only operates on `5` rather than `10`.

I guess we handle this, I can check further to confirm.

File: prql-compiler/src/sql/std.sql.prql [hunk 3]
```diff
@@ -56,29 +57,33 @@ let neg = l -> null
 let not = l -> null
 
 module postgres {
-    let regex_search = text pattern -> s"{text} ~ {pattern}"
+    let div_i = l r -> s"trunc({l} / {r})::bigint"
+    let round = n_digits column -> s"ROUND(({column})::numeric, {n_digits})"
+    let regex_search = text pattern -> s"({text} ~ {pattern})"
```
Comment (max-sixty) [inline prql-compiler/src/sql/std.sql.prql:62]: So good that we have the dialect modules now — adding something goes from "difficult rust" to "easy PRQL"!

File: prql-compiler/src/tests/test.rs [hunk 4]
```diff
@@ -1142,7 +1142,7 @@ fn test_nulls() {
     "###).unwrap()), @r###"
     SELECT
       *,
-      COALESCE(amount + 2, 15) AS amount
+      COALESCE(amount + 2, 3 * 5) AS amount
```
Comment (max-sixty) [inline prql-compiler/src/tests/test.rs:1145]: Out of interest, why did we remove these? (no normative view, just out of interest)

File: prql-compiler/tests/integration/queries/arithmetic.prql [hunk 5]
```diff
@@ -0,0 +1,27 @@
+[
+    { x_int =  13, x_float =  13.0, k_int =  5, k_float =  5.0 },
+    { x_int = -13, x_float = -13.0, k_int =  5, k_float =  5.0 },
+    { x_int =  13, x_float =  13.0, k_int = -5, k_float = -5.0 },
+    { x_int = -13, x_float = -13.0, k_int = -5, k_float = -5.0 },
```
Comment (max-sixty) [inline prql-compiler/tests/integration/queries/arithmetic.prql:5]: I'm guessing this isn't auto-formatted? It's beautiful! 

(Maybe that's a side project — align relation literals like this automatically!)

File: prql-compiler/src/sql/std.sql.prql (hunk 2 above)
Comment (max-sixty) [inline prql-compiler/src/sql/std.sql.prql:42]: @aljazerzen how would you feel about:
- Generic dialect is just `l / r`
- MySQL & MsSql use `s"({l} * 1.0 / {r})"`

That way, the default SQL is cleaner, and we keep the guarantees when a dialect is specified. I do still place some weight on having the introductory examples be really easy to read and compare.

The downside is that we don't guarantee the type of division on a generic dialect.

(I'm happy to make the change if you agree)

File: web/book/tests/snapshots/snapshot__queries__functions__piping__2.snap [hunk 6]
```diff
@@ -4,7 +4,7 @@ expression: "let fahrenheit_to_celsius = temp -> (temp - 32) / 1.8\nlet interp =
 ---
 SELECT
   *,
-  ((temp_c - 32) / 1.8 - 0) / 100 AS boiling_proportion
+  ((temp_c - 32 * 1.0 / 1.8) - 0 * 1.0 / 100 - 0) AS boiling_proportion
```
Comment (max-sixty) [inline web/book/tests/snapshots/snapshot__queries__functions__piping__2.snap:7]: Re the comment above re generic dialect — I do think this is now harder to read

File: prql-compiler/src/sql/std.sql.prql (hunk 2 above)
Comment (max-sixty) [reply prql-compiler/src/sql/std.sql.prql:42]: Actually this comes up in https://github.com/PRQL/prql/pull/2692

```
(started + unfinished * 1.0 / started)
# and
* 1.0 / 1 - (started ...
```

File: prql-compiler/src/sql/std.sql.prql (hunk 1 above)
Comment (max-sixty) [reply prql-compiler/src/sql/std.sql.prql]: Yes, this is correct — MySQL seems to have Decimal division rather than Float division...

File: prql-compiler/src/sql/std.sql.prql (hunk 2 above)
Comment (aljazerzen) [reply prql-compiler/src/sql/std.sql.prql:42]: We need more parenthesis!

Maybe move the "binding strength" calculation and parenthesis generation into here. Annotate the strength using some new annotation syntax?.

File: prql-compiler/src/tests/test.rs (hunk 4 above)
Comment (aljazerzen) [reply prql-compiler/src/tests/test.rs:1145]: I didn't feel like validating all of the evaluation rules I had. So I just deleted them.

File: prql-compiler/src/sql/std.sql.prql (hunk 2 above)
Comment (aljazerzen) [reply prql-compiler/src/sql/std.sql.prql:42]: I don't like that. Character `/` in PRQL means something else than `/` in SQL. I don't think it should be problematic to emit code that looks differently.

All of this would be way more ergonomic if we displayed query results in the book, instead of SQL.

Comment (max-sixty) [reply prql-compiler/src/sql/std.sql.prql:42]: > I don't like that. Character `/` in PRQL means something else than `/` in SQL.

I agree with the principle — if it's different then that's bad.
But it's only different in some minor dialects. My suggestion is to support those iff they're explicitly specified.

Another approach would be to default to DuckDB dialect (or Postgres, or Snowflake, i.e. one which has reasonable behavior here), rather than have the default try to cater to every dialect. Then we could use `/` safely.

> All of this would be way more ergonomic if we displayed query results in the book, instead of SQL.

Showing results would be useful too, but many people are much more used to SQL, and it's much clearer to them to see the actual SQL rather than see the results of a query on a table they're not familiar with.

I get that there's a tradeoff there — to what extent are we "producing SQL" vs "producing query results"? When I speak to people about PRQL, they do say they value the legible SQL output. And if I didn't know about PRQL, I would care about that — it's too difficult to have trust or learn the language otherwise. 

I think possibly we're on different sides of this discussion, I hope we can find a synthesis of our perspectives.

File: prql-compiler/src/tests/test.rs (hunk 4 above)
Comment (max-sixty) [reply prql-compiler/src/tests/test.rs:1145]: OK!

File: prql-compiler/src/sql/std.sql.prql (hunk 2 above)
Comment (max-sixty) [reply prql-compiler/src/sql/std.sql.prql:42]: I added one option in https://github.com/PRQL/prql/issues/2694...

Comment (aljazerzen) [reply prql-compiler/src/sql/std.sql.prql:42]: If we could be emitting simpler SQL that still passes the tests, we should definitely simply it. This is not hard now.

I don't strongly feel we should switch to showing query results instead of SQL, just a suggestion.

But I do feel strongly that if we decide on different semantics than SQL, we should compile to something that does implement those semantics correctly, in all edge cases. Otherwise our argument that "PRQL can be used with any SQL database" is simply not true.

We could present this as "PRQL handles edge cases for you".

File: prql-compiler/tests/integration/docker-compose.yml [hunk 7]
```diff
@@ -1,3 +1,4 @@
+version: "2.4"
```
Comment (eitsupi) [inline prql-compiler/tests/integration/docker-compose.yml:1]: This line seems added by accident?

File: prql-compiler/src/sql/std.sql.prql [hunk 8]
```diff
@@ -56,29 +57,33 @@ let neg = l -> null
...
 
 module bigquery {
     let regex_search = text pattern -> s"REGEXP_CONTAINS({text}, {pattern})"
 }
 
 module duckdb {
-    let regex_search = text pattern -> s"REGEXP_MATCHES({text}, {pattern})"
-}
+    # TODO: uncomment when updated on latest DuckDB version
+    # let div_i = l r -> s"({l} // {r})"
```
Comment (eitsupi) [inline prql-compiler/src/sql/std.sql.prql:86]: DuckDB's `//` does not work for floats, so I think we can't use this.
duckdb/duckdb#7558

File: prql-compiler/tests/integration/docker-compose.yml (hunk 7 above)
Comment (aljazerzen) [reply prql-compiler/tests/integration/docker-compose.yml:1]: Nope, it's intentional. 

Docker-compose version 3.x was rejecting the file, saying mssql contains entries that are not allowed.

Comment (eitsupi) [reply prql-compiler/tests/integration/docker-compose.yml:1]: That is weird. Do you know what version of docker compose you are using?

- Currently the latest docker compose is 2.18.1 (See https://github.com/docker/compose)
- docker compose 2.X is based on the compose spec, and `version` is deprecated in the compose spec. https://github.com/compose-spec/compose-spec/blob/master/spec.md#version-top-level-element

Comment (aljazerzen) [reply prql-compiler/tests/integration/docker-compose.yml:1]:
```
$ docker-compose --version
docker-compose version 1.25.0, build unknown
```

File: prql-compiler/src/sql/std.sql.prql (hunk 2 above)
Comment (max-sixty) [reply prql-compiler/src/sql/std.sql.prql:42]: I think we agree on the principles:
- PRQL should be correct (not approximately correct)
- We should have legible SQL

...and here they're in conflict. 

I probably weigh Legibility more highly than you?

Is there a synthesis of these tradeoffs where we can do both? 
- Default compile to DuckDB dialect?
- Default compile to DuckDB on the website & playground?
- Have a small list of supported dialects that work with the generic dialect, and others need to explicitly specify the dialect

WDYT?

File: prql-compiler/tests/integration/docker-compose.yml (hunk 7 above)
Comment (eitsupi) [reply prql-compiler/tests/integration/docker-compose.yml:1]: > $ docker-compose --version
> docker-compose version 1.25.0, build unknown

Oh, your docker-compose version is outdated. compose version 1 is no longer supported, and compose version 2 is widely used these days.
compose v1 is written in Python, and compose v2 is fully rewritten in Go as Docker plugin.
Please uninstall compose v1 and install v2.
https://docs.docker.com/compose/install/
